File: mirror/tier2-php/showcase/ir_canonical.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: IR & Determinism
// Module: ir_canonical.php

function canonical_form($expr) {
    if (is_array($expr)) {
        if (($expr['op'] ?? null) === 'add') {
            $left = canonical_form($expr['left']);
            $right = canonical_form($expr['right']);
            
            if (is_numeric($left) && is_numeric($right)) {
                return $left + $right;
            }
            
            return ['op' => 'add', 'left' => $left, 'right' => $right];
        } elseif (($expr['op'] ?? null) === 'mul') {
            $left = canonical_form($expr['left']);
            $right = canonical_form($expr['right']);
            
            if (is_numeric($left) && is_numeric($right)) {
                return $left * $right;
            }
            
            return ['op' => 'mul', 'left' => $left, 'right' => $right];
        }
    }
    
    return $expr;
}

$expr1 = ['op' => 'add', 'left' => 2, 'right' => 3];
$expr2 = ['op' => 'mul', 'left' => 4, 'right' => 5];
$expr3 = ['op' => 'add', 'left' => ['op' => 'mul', 'left' => 2, 'right' => 3], 'right' => 4];

$result = [
    'canon_add' => canonical_form($expr1),
    'canon_mul' => canonical_form($expr2),
    'canon_nested' => canonical_form($expr3),
    'canonicalization_active' => true,
];

echo json_encode($result);
?>

File: mirror/tier2-dart/showcase/ir_canonical.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: IR & Determinism
// Module: ir_canonical.dart

dynamic canonicalForm(dynamic expr) {
  if (expr is Map) {
    if (expr['op'] == 'add') {
      final left = canonicalForm(expr['left']);
      final right = canonicalForm(expr['right']);
      
      if (left is num && right is num) {
        return left + right;
      }
      
      return {'op': 'add', 'left': left, 'right': right};
    } else if (expr['op'] == 'mul') {
      final left = canonicalForm(expr['left']);
      final right = canonicalForm(expr['right']);
      
      if (left is num && right is num) {
        return left * right;
      }
      
      return {'op': 'mul', 'left': left, 'right': right};
    }
  }
  
  return expr;
}

void main() {
  final expr1 = {'op': 'add', 'left': 2, 'right': 3};
  final expr2 = {'op': 'mul', 'left': 4, 'right': 5};
  final expr3 = {
    'op': 'add',
    'left': {'op': 'mul', 'left': 2, 'right': 3},
    'right': 4,
  };
  
  final result = <String, dynamic>{
    'canonAdd': canonicalForm(expr1),
    'canonMul': canonicalForm(expr2),
    'canonNested': canonicalForm(expr3),
    'canonicalizationActive': true,
  };
  print(result);
}
